base/multidimensional.jl

```diff
@@ -334,9 +334,14 @@ module IteratorsMD
         iterfirst, iterfirst
     end
     @inline function iterate(iter::CartesianIndices, state)
-        nextstate = CartesianIndex(inc(state.I, first(iter).I, last(iter).I))
-        nextstate.I[end] > last(iter.indices[end]) && return nothing
-        nextstate, nextstate
+        # If we increment before the condition check, we run the
+        # risk of an integer overflow.
+        if state.I[end] < last(iter.indices[end])
```

Comment:
perhaps a stupid question, but doesn't this do one fewer iterations than before?

Reply:
If so then we definitely need a test to catch that 😁

Reply:
xD, indeed for the ND-case this would drop the last few iterations. Added a test and @YingboMa fixed the logic.